### PR TITLE
Allow compiling to JS on-the-fly

### LINKED ISSUE #106: auto-compile TS -> JS
Currently this works: https://deno.land/std/http/server.ts

It would also be nice if this worked too: https://deno.land/std/http/server.js

Where the server magically runs tsc on the file (without type checks) and produces the corresponding JS.

This would allow browsers to use deno modules.

### PR DESCRIPTION
**Pull request by j-f1**
Fixes #106.

@ry suggested this use [the same TSConfig file that Deno uses](https://github.com/denoland/deno/blob/240ca25617bb57e4071193dbc040ab03a27f3a6c/js/compiler.ts#L151-L161). Here’s a quick run-through of those options:

- `allowJs` ✅
- `checkJs` 🚫 we don’t want to produce any errors — we don’t judge, we transpile.
- `esModuleInterop` ✅
- `module` ✅
- `outDir` 🚫 we aren’t writing to a file
- `resolveJsonModule` 🚫 we don’t handle any imports
- ~`sourceMap`~ `inlineSourceMap` ✅ [suggested](https://github.com/denoland/registry/issues/39#issuecomment-508607314) by @kitsonk
- `stripComments` ✅
- `target` ✅

I think the only issue with using this vs. native TS support is that it doesn’t support imported `const enum`s. There isn’t really a good way to fix this, so we’d be better off putting something in the style guide that advises against them, since they don’t serve any special purpose.

Screen Shot 2019-07-12 at 2 25 23 PM

### CODE REVIEW

File: src/transpile_code.js [hunk 1]
```diff
@@ -0,0 +1,78 @@
+const ts = require("typescript");
+
+const response = require("./response");
+
+module.exports = function compile(pathname, content) {
```
Comment (ry) [inline src/transpile_code.js]: I think we should have a unit test for this

File: src/transpile_code.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+          ts.updateSourceFileNode(
+            sourceFile,
+            transformImports(sourceFile.statements, pathname)
+          )
+      ]
+    }
+  });
+};
+
+function transformImports(statements, pathname) {
```
Comment (ry) [inline src/transpile_code.js]: I think a unit test for this too...

File: src/transpile_code.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+          node.modifiers,
+          node.name,
+          ref
+        );
+    }
+    return node;
+  });
+}
+
+function transformModuleSpecifier(specifier, pathname) {
```
Comment (ry) [inline src/transpile_code.js]: test here too

**Comment by j-f1** [comment]
> tsconfig values you suggested sound good

Should I strip comments or keep them?

**Comment by ry** [comment]
I guess strip

File: src/transpile_code.js (hunk 2 above)
Comment (j-f1) [reply src/transpile_code.js]: Is it OK if I just test it as a part of the overall transpile process?

Comment (ry) [reply src/transpile_code.js]: Not sure - leave it out for now and do a unit test for the compile function. 

I feel like having small unit tests for things like compilation is important, but we can do it incrementally.

File: src/test/transpile_snapshot.js [hunk 4]
```diff
@@ -0,0 +1,5 @@
+import { foo2 } from "./type_import.ts?transpile";
```
Comment (ry) [inline src/test/transpile_snapshot.js]: change the src/test directory to src/testdata that makes it more clear that this is "data" rather than infrastructure code in the test suite.

File: src/test/transpile.js [hunk 5]
```diff
@@ -0,0 +1,40 @@
+const assert = require("assert");
```
Comment (ry) [inline src/test/transpile.js]: Rename this file to `src/transpile_code_test.js` as it contains actual unit tests.

File: src/test/server.js [hunk 6]
```diff
@@ -31,49 +32,40 @@ async function tests() {
...
   assert.equal(counter, 6);
   assert.equal(response.status, "200");
 
-  event = require("./testdata/req7.json");
+  event = require("./data/req7.json");
   await lambdaHandler(event, context, callback);
   assert.equal(counter, 7);
   assert.equal(response.status, "200");
 
-  event = require("./testdata/req8.json");
```
Comment (ry) [inline src/test/server.js]: I don't like this rename, I think it should remain in src/testdata for clarity.

**Comment by ry** [comment]
Could you add a link on HTML source view page to the `?transpile` version? Otherwise it will be difficult to discover.

File: package.json [hunk 7]
```diff
@@ -0,0 +1,21 @@
+{
```
Comment (ry) [inline package.json]: Since you're moving package.json to the root (which seems reasonable), please remove the "cd src" from this line too
https://github.com/denoland/registry/blob/2ab2e6e9a338cc402a8747ed05f9ae1d03a500ba/deploy.sh#L18

**Comment by j-f1** [comment]
What about moving the tests to a separate `tests/` directory in the root of the repo? This increases flatness.

**Comment by ry** [comment]
@j-f1 The "fixtures" can be put into `//testdata`, but the test code itself should be a neighbor of the code it is testing. See for example the organization of `std/encoding` https://github.com/denoland/deno_std/tree/master/encoding

**Comment by ry** [comment]
Nice tests! Looking good.

What about the situation where there's a non-TS file and someone adds "?transpile" ... it seems like that should result in a 500 HTTP error? Example: http://localhost:4000/x/install/install.ps1?transpile

**Comment by ry** [comment]
Rather than `?transpile` what do you think about `?js`

It's shorter and more descriptive.

File: src/render_code.js [hunk 8]
```diff
@@ -59,6 +59,9 @@ module.exports = function renderCode(pathname, code, repo) {
       </head>
       <body>
         <a href="${repo}">View repository on GitHub</a>
+        ${opts.compiled
+          ? `<br><br><em>This file has been compiled to JS. <a href="${url}">View the original version here</a>.</em>`
+          : ""}
```
Comment (ry) [inline src/render_code.js]: Could you add a link on typescript source pages to the `?transpile` version? 

This will help discovery of this feature.

**Comment by ry** [comment]
I have trouble deploying this patch to AWS because you've moved the package.json out of the src folder. I'm not sure how to solve this - and these changes are unrelated to compiling on-the-fly - so let's undo those for now.

File: deploy.sh [hunk 9]
```diff
@@ -15,7 +15,7 @@ print_errors() {
   false
 }
 
-(cd src && npm install)
+npm install
```
Comment (ry) [inline deploy.sh]: Undo this change too

Comment (ry) [reply deploy.sh]: I will experiment with moving package.json to the root separately. Sorry about that.

Comment (j-f1) [reply deploy.sh]: No worries!

**Comment by ry** [comment]
I've deployed it.. Check it out https://deno.land/std/io/bufio.ts?js

I'm just going to let this run for a while before landing - to check if there are any problems.
